services/storage/store.go

```diff
@@ -312,60 +353,96 @@ func (s *Store) TagValues(ctx context.Context, req *datatypes.TagValuesRequest)
 		return nil, err
 	}
 
-	database, rp, start, end, err := s.validateArgs(source.Database, source.RetentionPolicy, req.Range.Start, req.Range.End)
+	db, rp, start, end, err := s.validateArgs(source.Database, source.RetentionPolicy, req.Range.Start, req.Range.End)
 	if err != nil {
 		return nil, err
 	}
 
-	shardIDs, err := s.findShardIDs(database, rp, false, start, end)
-	if err != nil {
-		return nil, err
-	}
-	if len(shardIDs) == 0 {
-		return cursors.NewStringSliceIterator(nil), nil
-	}
-
-	var expr influxql.Expr
+	var influxqlPred influxql.Expr
 	if root := req.Predicate.GetRoot(); root != nil {
 		var err error
-		expr, err = reads.NodeToExpr(root, measurementRemap)
+		influxqlPred, err = reads.NodeToExpr(root, measurementRemap)
 		if err != nil {
 			return nil, err
 		}
 
-		if found := reads.HasFieldValueKey(expr); found {
+		if found := reads.HasFieldValueKey(influxqlPred); found {
 			return nil, errors.New("field values unsupported")
 		}
-		expr = influxql.Reduce(influxql.CloneExpr(expr), nil)
-		if reads.IsTrueBooleanLiteral(expr) {
-			expr = nil
+
+		influxqlPred = influxql.Reduce(influxql.CloneExpr(influxqlPred), nil)
+		if reads.IsTrueBooleanLiteral(influxqlPred) {
+			influxqlPred = nil
+		}
+	}
+
+	mqAttrs := &metaqueryAttributes{
+		db:    db,
+		rp:    rp,
+		start: start,
+		end:   end,
+		pred:  influxqlPred,
+	}
+
+	tagKey, ok := measurementRemap[req.TagKey]
+	if !ok {
+		tagKey = req.TagKey
+	}
+
+	// Getting values of _measurement or _field are handled specially
+	switch tagKey {
+	case "_name":
+		return s.MeasurementNames(ctx, mqAttrs)
+
+	case "_field":
+		return s.measurementFields(ctx, mqAttrs)
+	}
+
+	return s.tagValues(ctx, mqAttrs, tagKey)
+}
+
+func (s *Store) tagValues(ctx context.Context, mqAttrs *metaqueryAttributes, tagKey string) (cursors.StringIterator, error) {
+	// If there are any references to _field, we need to use the slow path
+	// since we cannot rely on the index alone.
+	if mqAttrs.pred != nil {
+		if hasFieldKey := reads.ExprHasKey(mqAttrs.pred, fieldKey); hasFieldKey {
+			return s.tagValuesSlow(ctx, mqAttrs, tagKey)
```

Comment:
This also needs new tests in `show tag values`

Reply:
Confirmed that InfluxQL is unaffected by this change, adding a `panic` doesn't cause any existing tests to fail